File: sdk/oneagent/java/NeuralOpsAgent.java
package com.neuralops.oneagent;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.json.JSONObject;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/** Lightweight HTTP auto-instrumentation for Java services. */
public final class NeuralOpsAgent {
    private NeuralOpsAgent() {}

    public static OkHttpClient.Builder instrument(OkHttpClient.Builder builder, Config config) {
        return builder.addInterceptor(new SpanInterceptor(config));
    }

    public static final class Config {
        public String serviceName = System.getenv().getOrDefault("NEURALOPS_SERVICE", "java-service");
        public String ingestUrl = System.getenv().getOrDefault("NEURALOPS_INGEST_URL", "http://localhost:8080/api/v1/apm/profiles");
        public String tenantId = System.getenv().getOrDefault("NEURALOPS_TENANT_ID", "default");
    }

    static final class SpanInterceptor implements Interceptor {
        private final Config config;
        SpanInterceptor(Config config) { this.config = config; }

        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            long start = System.nanoTime();
            Response response = chain.proceed(request);
            try {
                JSONObject body = new JSONObject();
                body.put("functionName", request.method() + " " + request.url().encodedPath());
                body.put("filePath", "okhttp3");
                body.put("lineNo", response.code());
                body.put("selfTimeMs", TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start));
                body.put("sampleCount", 1);
                body.put("service", config.serviceName);
                okhttp3.Request emit = new okhttp3.Request.Builder()
                    .url(config.ingestUrl)
                    .post(okhttp3.RequestBody.create(body.toString(), okhttp3.MediaType.parse("application/json")))
                    .header("X-Tenant-ID", config.tenantId)
                    .build();
                new OkHttpClient().newCall(emit).execute().close();
            } catch (Exception ignored) { }
            return response;
        }
    }
}

File: sdk/oneagent/dotnet/NeuralOpsAgent.cs
using System.Diagnostics;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace NeuralOps.OneAgent;

/// <summary>HTTP auto-instrumentation handler for .NET services.</summary>
public sealed class NeuralOpsHandler : DelegatingHandler
{
    private readonly NeuralOpsConfig _config;

    public NeuralOpsHandler(NeuralOpsConfig config, HttpMessageHandler? inner = null)
        : base(inner ?? new HttpClientHandler())
    {
        _config = config;
    }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var sw = Stopwatch.StartNew();
        var response = await base.SendAsync(request, cancellationToken);
        sw.Stop();
        _ = EmitSpanAsync(request, response, sw.ElapsedMilliseconds);
        return response;
    }

    private async Task EmitSpanAsync(HttpRequestMessage request, HttpResponseMessage response, long durationMs)
    {
        try
        {
            using var client = new HttpClient();
            var payload = JsonSerializer.Serialize(new
            {
                functionName = $"{request.Method} {request.RequestUri?.AbsolutePath}",
                filePath = "System.Net.Http",
                lineNo = (int)response.StatusCode,
                selfTimeMs = durationMs,
                sampleCount = 1,
                service = _config.ServiceName,
            });
            using var content = new StringContent(payload, Encoding.UTF8, "application/json");
            content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
            var req = new HttpRequestMessage(HttpMethod.Post, _config.IngestUrl) { Content = content };
            req.Headers.Add("X-Tenant-ID", _config.TenantId);
            await client.SendAsync(req);
        }
        catch
        {
            // best-effort
        }
    }
}

public sealed class NeuralOpsConfig
{
    public string ServiceName { get; init; } = Environment.GetEnvironmentVariable("NEURALOPS_SERVICE") ?? "dotnet-service";
    public string IngestUrl { get; init; } = Environment.GetEnvironmentVariable("NEURALOPS_INGEST_URL") ?? "http://localhost:8080/api/v1/apm/profiles";
    public string TenantId { get; init; } = Environment.GetEnvironmentVariable("NEURALOPS_TENANT_ID") ?? "default";
}
